src/language_features/kak_lsp_custom.rs

```diff
@@ -0,0 +1,318 @@
+/// current cursor location?
+fn exceeds(
+    symbol_position: KakounePosition,
+    symbol_kind: SymbolKind,
+    cur_position: KakounePosition,
+    search_next: bool,
+    expected_kind: &str,
+) -> bool {
+    // expected kind is the symbol for which the user is searching for
+    // e.g. `Function`, `Constructor` etc.
+    if !expected_kind.is_empty() && format!("{:?}", symbol_kind) != expected_kind {
+        return false;
+    }
+
+    // If searching forwards, the first element that has a greater line/column
+    // combination
+    // If searching backwards, the first element that has a smaller line/column
+    // combination
+    if search_next {
+        symbol_position > cur_position
+    } else {
+        symbol_position < cur_position
+    }
+}
+
+/// Some languages modify the name of the function. This function normalizes
+/// them so that they can be found in the document.
+fn process_name<'a>(ctx: &Context, name: &'a str) -> &'a str {
+    if ctx.language_id == "erlang" {
+        // In erlang the arity of the function is added to the function name
+        // e.g. `foo` function may be named something like `foo/3`
+        name.split('/').next().unwrap()
+    } else {
+        name
+    }
+}
+
+fn special_case_position(ctx: &Context, pos: Position) -> Option<Position> {
+    // In Elixir we don't need to search forward.
+    // The returned position is basically the selection range start.
+    if ctx.language_id == "elixir" {
+        Some(pos)
+    } else {
+        None
+    }
+}
+
+/// Gets (filename, kakoune position, name) of the next/previous
+/// SymbolInformation symbol in the document
+fn get_next_or_prev_symbol_information_details(
+    items: Vec<SymbolInformation>,
+    params: NextOrPrevSymbolParams,
+    ctx: &Context,
+) -> Option<(String, KakounePosition, String, SymbolKind)> {
+    // Setup an iterator dependending on whether we are searching forwards or backwards
+    let it: Box<dyn Iterator<Item = &SymbolInformation>> = if params.search_next {
+        Box::new(items.iter())
+    } else {
+        Box::new(items.iter().rev())
+    };
+
+    for SymbolInformation {
+        location,
+        kind,
+        name,
+        ..
+    } in it
+    {
+        let filename_path = location.uri.to_file_path().unwrap();
+        let filename = filename_path.to_str().unwrap().to_owned();
+
+        let symbol_position = get_symbol_hover_pos(&filename, ctx, location, name);
+
+        if exceeds(
+            symbol_position,
+            *kind,
+            params.position,
+            params.search_next,
+            &params.symbol_kind,
+        ) {
+            return Some((filename, symbol_position, name.to_owned(), *kind));
+        }
+    }
+
+    None
+}
+
+fn get_symbol_hover_pos(
+    filename: &String,
+    ctx: &Context,
+    location: &Location,
+    name: &String,
+) -> KakounePosition {
+    // We need to handle some language servers specially
```

Comment:
If I open our `src/controller.rs` using rust-analyzer, scroll to the bottom and
run `lsp-goto-previous-function`, then the cursor ends up inside the docstring,
in a line that reads

	/// by the incoming request. `ensure_did_open` tries to sneak in `textDocument/didOpen` request for

This is because we use `find_name_in_file` to "correct" the position that the language server gives us - the start of the doc comment. We want to the "meat", the "fn ensure_did_open". 
So `find_name_in_file` is a workaround for missing information in LSP. I guess it's fine as long as it doesn't make realistic cases worse. We can probably ignore the case of `ensure_did_open` since it's rare? I'm not sure, I usually work with more concise  doc comments, if any.

So I guess `location.range.start` is not good enough. I guess, because of doc comments, as mentioned above.
Does `selection_range` work? Ideally we could drop language-specific workarounds.

Reply:
> But most LSPs invoked by `kak-lsp` seem to return `struct SymbolInformation` on the `textDocument/documentSymbol` request only. 
		
thanks for this information, makes sense

I guess the workarounds are okay because they only apply to servers that send SymbolInformation, and failures should be rare.
It might be nice to allow users to add their own custom position-computation logic in kakscript, but that can be done later.

> How can we start using `selection_range` ? Is there some setting on the kak-lsp side that will start allowing LSPs to send us their results as `DocumentSymbol`?

right, I'll try to check now. I mostly care about gopls, clangd and rls/rust-analyzer, I'll check those

Reading gopls sources got me there quickly

```go
	// TODO: Remove this once the lsp deprecates SymbolInformation.
	symbols := make([]interface{}, len(docSymbols))
	for i, s := range docSymbols {
		if snapshot.View().Options().HierarchicalDocumentSymbolSupport {
			symbols[i] = s
			continue
		}
		// If the client does not support hierarchical document symbols, then
		// we need to be backwards compatible for now and return SymbolInformation.
		symbols[i] = protocol.SymbolInformation{...}
```

so it's just

```
diff --git a/src/general.rs b/src/general.rs
index fbd26eb..f5a742c 100644
--- a/src/general.rs
+++ b/src/general.rs
@@ -163,7 +163,7 @@ pub fn initialize(root_path: &str, meta: EditorMeta, ctx: &mut Context) {
                 document_symbol: Some(DocumentSymbolClientCapabilities {
                     dynamic_registration: Some(false),
                     symbol_kind: None,
-                    hierarchical_document_symbol_support: None,
+                    hierarchical_document_symbol_support: Some(true),
                     tag_support: None,
                 }),
                 formatting: Some(DynamicRegistrationClientCapabilities {
```

then it magically works. Very nice that we finally notice this, now we can also tweak the standard document `lsp-document-symbol` to maybe use the selection_range.